### PR TITLE
Beam/Pulse lasers

### PR DESCRIPTION
**Pull request by fluffyfreak**
A Beam/Pulse laser branch, ~~probably not going to look/work like this in the end but it's been loitering on my fork and needs pushing here so people can discuss.~~

Decided to go with this with a view to improving it later. The idea being that people can start playing with it and giving feedback.

There's some balancing that needs doing before it's truly ready for merging, need feedback and thoughts on:
- Damage done per shot
- Heat generated per shot
- Cost of the beam weapons

Thoughts and opinions welcome.

### CODE REVIEW

**Comment by nozmajner** [comment]
Cool! Just tried it (still got my ass handed to me though)

**Comment by richardpl** [comment]
Just tested, nice work.

I guess the damage should be at least 1.5x time bigger.
The heat generation looks almost fine, perhaps more the stronger laser is.
The weakest laser could never reach heat status, but it should if constantly on.
Price should be IMO twice than currently is.
Also needs to add stronger lasers that weight more.

**Comment by fluffyfreak** [comment]
Good suggestions I'll tweak the price, damage and I've bumped up the mass a bit more so it seems more in line with the pulse weapons.

**Comment by impaktor** [comment]
Oh, so this isn't WIP anymore? :fireworks:

**Comment by impaktor** [comment]
> Whole point of beam lasers is that they should heat up faster than cannon ones.

I think the point of beam lasers -- and the reason it's been requested so many times -- is that players want something where the projectiles are way way faster then the current weapons, which are very difficult to hit a moving target with. 

Beam weapons, with instantaneous speed, worked nicely in Frontier, so I hope this could work in a similar way.

**Comment by richardpl** [comment]
I tested it and it works as intended, just needs some more heat generation.

**Comment by impaktor** [comment]
Also, this needs some rebasing. I can do that, but there looks to be merge conflicts atm, according to github.

**Comment by fluffyfreak** [comment]
@impaktor I fixed the merge conflicts last night, they were due to transfiex PR.

**Comment by fluffyfreak** [comment]
Ok, I think that's all the changes needed. Beam weapons now heat things up twice as fast and this can now also be controlled from equipment script with new optional parameters `heatrate` and `coolrate`

**Comment by richardpl** [comment]
Ready for merge, feel free to add heavier variants of beam weapons too.

**Comment by impaktor** [comment]
I'm currently a bit AFK, but please don't merge without a rebase; alternatively, wait for me to merge.

**Comment by richardpl** [comment]
@impaktor why it needs rebase? There are no conflicts.

**Comment by fluffyfreak** [comment]
@richardpl because of the ugly commit history, merges, things added, removed then added again.

**Comment by impaktor** [comment]
I'll get to this next week, a lot on my plate currently.

**Comment by fluffyfreak** [comment]
@impaktor no worries, and no rush as usual :)

File: data/libs/Equipment.lua [hunk 1]
```diff
@@ -886,70 +886,94 @@ laser.pulsecannon_1mw = LaserType.New({
 	l10n_key="PULSECANNON_1MW", price=600, capabilities={mass=1},
 	slots = {"laser_front", "laser_rear"}, laser_stats = {
 		lifespan=8, speed=1000, damage=1000, rechargeTime=0.25, length=30,
-		width=5, dual=0, mining=0, rgba_r = 255, rgba_g = 51, rgba_b = 51, rgba_a = 255
+		width=5, beam=0, dual=0, mining=0, rgba_r = 255, rgba_g = 51, rgba_b = 51, rgba_a = 255
```
Comment (impaktor) [inline data/libs/Equipment.lua]: Looks like you're using `beam` as a boolean, and not an integer? 

Just a comment, no action needed.

**Comment by impaktor** [comment]
Wasn't this how the frontier version sounded like (very approximate)?
https://freesound.org/people/senitiel/sounds/131431/

Not too bad, perhaps?
https://freesound.org/people/bubaproducer/sounds/151022/

Some suggestions:
https://freesound.org/people/mparsons99/sounds/89490/

This, minus the warm-up sound, if anyone wants to play around with audacity:
https://freesound.org/people/tjcason/sounds/390474/

Maybe this, except the loading sound in the start:
http://soundbible.com/1774-Laser-Machine-Gun.html

Or this? :wink: 
http://soundbible.com/1799-Ima-Firin-My-Laser.html

This could be a place for further research:
https://medium.com/@PeakselDev/top-17-websites-with-free-sound-effects-and-music-for-your-mobile-app-f2ef49cbf9eb

**Comment by WKFO** [comment]
We... we already have a beam laser sound in game files.
But it is fine, we can change it!

**Comment by WKFO** [comment]
I just tested. Unfortunately that doesn't do very well. Beam laser fires so rapidly the sound just overlaps. We'll have to either script the thing to not play the sound if it is already playing, or get a new sound.

So far that's the closest thing I have to a laser pulse:
https://soundcloud.com/homeworld-kerbin/beam-laser-pulse

**Comment by WKFO** [comment]
Also, while testing the dual beam laser on Natrix, it is a little off. I dunno if it happens on other vessels. Maybe we can do sth about it?

**Comment by impaktor** [comment]
> Beam laser fires so rapidly the sound just overlaps (and keeps playing a while after you stop shooting)

I wonder if this perhaps can be solved in code?

**Comment by WKFO** [comment]
Most probably.

**Comment by fluffyfreak** [comment]
I admit that I hadn't tested with sound! Stupid of me, yes this will need fixing in code I think.

**Comment by fluffyfreak** [comment]
That pulse laser sound is really naff. It needs to be sharper, to sound more like a laser rather than the wierd pusling noise that's currently there.

The trouble with having multiple sounds is that the beam can be very active for a very short amount of time.

**Comment by WKFO** [comment]
We aren't using `sounds/ship/beam_laser.ogg`?

If we wanna go with the pulse sounds, impaktor has bunch of ideas, or perhaps sth more like: 
https://soundcloud.com/homeworld-kerbin/laser3b

**Comment by fluffyfreak** [comment]
Dammit, i meant the beam laser sound, have updated my comment.

**Comment by fluffyfreak** [comment]
Ok, beam laser sounds now work. Had to do some strange stuff to the sound system... that's some scary code.

**Comment by impaktor** [comment]
Sound seems to still indicate it's shooting once it's over-heated and not shooting.

**Comment by fluffyfreak** [comment]
@impaktor I am not able to repro' the continuous firing, however I can try to prevent it by disabling the "is firing" flag on overheating

**Comment by impaktor** [comment]
I'll try it in the comming days. (At work now). Looks like github is saying there be conflicts?

**Comment by fluffyfreak** [comment]
It says it has no conflicts for me?

**Comment by impaktor** [comment]
OK, I've tried cleaning this up, but it's hopeless. Cherry-picking the commits to a new branch, I get conflicts I'm not too keen on having to deal with. Trying to do a diff between branches, and then applying the changes to a fresh branch fails, for unknown reason. I'm doing something wrong, I assume.

I'm abandoning my atempts for today. @laarmen do you have any git wizardy how to clean up this PR?

@fluffyfreak any reason you keep merging in other branches in this? (master, specifically). Were they to resolve conflicts?

I'd like to see what ecraven says when he gets back (in one week?)

**Comment by johnbartholomew** [comment]
Rebased fluffyfreak/pewpew available at https://github.com/johnbartholomew/pioneer/tree/clean-pewpew

fluffyfreak/pewpew with master merged in again (for comparison with the rebased version) available at:
https://github.com/johnbartholomew/pioneer/tree/reverse-pewpew

They aren't identical, but I believe the differences are intended/better on the rebased version. Compare with:
git diff johnbartholomew/reverse-pewpew johnbartholomew/clean-pewpew

**Comment by laarmen** [comment]
Usually when it seems hopeless, I just do a soft reset on the current
master, and re-commit the code using add -p to only take the relevant bits.
So, no, no wizardry.

**Comment by impaktor** [comment]
Thanks @jpab for rebasing it. That made it so I could squash some stuff, like put new texture in same commit that use them, and add linux makefile-build fix to same commit that adds the new class and VS update. Now I can sleep.

@fluffyfreak big cheers for giving us beam weapons! :fireworks:

**Pull request merged**